libdnf5-cli/output/search.cpp

```diff
@@ -97,19 +98,64 @@ void print_search_results(const SearchResults & results) {
+    std::unique_ptr<libscols_table, decltype(&scols_unref_table)> table(scols_new_table(), &scols_unref_table);
+
+    if (!libdnf5::cli::tty::is_interactive()) {
+        // Do not hard-code 80 as non-interactive screen width. Let libdnf5::cli::tty to decide
+        auto screen_width = size_t(libdnf5::cli::tty::get_width());
+        scols_table_set_termwidth(table.get(), screen_width);
+        // The below is necessary to make the libsmartcols' truncation work with non-interactive terminal
+        scols_table_set_termforce(table.get(), SCOLS_TERMFORCE_ALWAYS);
+    }
+
+    const std::vector<std::tuple<const char *, double, int>> columns = {
+        {_("Package"), 0.15, 0}, {_("Description"), 0.7, SCOLS_FL_TRUNC}, {_("Matched fields"), 0.15, 0}};
+
+    for (auto [name, whint, flags] : columns) {
+        scols_table_new_column(table.get(), name, whint, flags);
+    }
 
+    for (auto const & result : results.group_results) {
+        auto matched_fields = construct_keys_string(result.matched_keys);
         for (auto const & package : result.matched_packages.packages) {
-            std::cout << " ";
+            struct libscols_line * ln = scols_table_new_line(table.get(), NULL);
+            std::string name;
             if (results.options.show_duplicates) {
-                std::cout << highlight(package.get_full_nevra());
+                name = package.get_full_nevra();
             } else {
-                std::cout << highlight(package.get_name()) << "." << package.get_arch();
+                name = package.get_name() + "." + package.get_arch();
             }
-            std::cout << ": " << highlight(package.get_summary()) << std::endl;
+            scols_line_set_data(ln, 0, name.c_str());
+            // Regretfully we can not use the highlight for the below because of truncation
+            scols_line_set_data(ln, 1, package.get_summary().c_str());
+            scols_line_set_data(ln, 2, matched_fields.c_str());
         }
     }
+
+    char * output_buffer;
+    size_t output_size;
+    FILE * output_stream;
+    output_stream = open_memstream(&output_buffer, &output_size);
+    scols_table_set_stream(table.get(), output_stream);
+    scols_print_table(table.get());
+    char line[4096];  // The usual maximum length of terminal line
+    // Printing table header in bold text
+    if (!fgets(line, sizeof(line), output_stream)) {
+        std::cout << _("Unable to retrieve search results.") << std::endl;
+        return;
+    }
+    std::cout << bold << std::string(line) << reset;
+    // Printing the rest of table with highlighting (only for first 2 columns)
+    auto pkg_n_desc_width = scols_column_get_width(scols_table_get_column(table.get(), 0));
+    pkg_n_desc_width += scols_column_get_width(scols_table_get_column(table.get(), 1));
+    pkg_n_desc_width += 2;  // extra space between columns
+    while (fgets(line, sizeof(line), output_stream)) {
```

Comment:
Same as above, use getline() in your read loop rather than fgets().  output_stream needs to be initialized to NULL.